### PR TITLE
Redoing Execution Plan Comparison Editor

### PR DESCRIPTION
**Pull request by aasimkhan30**
Icon for Plan Comparison Tab:
image

Icon for Execution Plan tab:
image

Entry points:
Using existing execution plan
image
image
Comparison editor opens with top graph already loaded:
image

Using command:
image
image

Toolbar for editor:
image

Dropdown for multiple plans:
image

Toggle Orientation:
Vertical:
image
azuredatastudio_whRdmiBJWH

Horizontal:
image
azuredatastudio_YhtYEuq4gd

Properties:
image
azuredatastudio_YDKWKkFCGw

Dark Mode:
imageHigh Contrast:
image

### CODE REVIEW

**Review by erinstellato-ms — COMMENTED** [review]
I think this looks good so far.  My only feedback is that when you have zero or one plan loaded, it would be nice it we could draw the user's attention to the "Add execution button".  Is it possible to bold that text?  I see that it becomes disabled when there are two plans selected, which is great.

I think that, if no plans are select, the Zoom buttons, and Toggle Orientation, should also be disabled.  And if only one plan is selected, what happens if you click on Toggle Orientation or Properties?  Do those need to be disabled in the one plan scenario?

Lastly, I know that for testing you are using different queries for comparison, I'd like to understand how this works.  Meaning: if I have 5 queries and their plans in one plan file, with that drop down, and I select ANY query for the top plan and any query for the bottom plan?

**Comment by aasimkhan30** [comment]
If you want to open compare two plans in the same file, you will have to add the file twice and then compare the query plans by selecting the right ones in both dropdowns.

**Comment by erinstellato-ms** [comment]
> If you want to open compare two plans in the same file, you will have to add the file twice and then compare the query plans by selecting the right ones in both dropdowns.

Ah, that wasn't my question.  Let's say I have these 3 queries in a batch that I run together, and so there is one .sqlplan file with the plan for all three queries:

SELECT * FROM dbo.Orders
SELECT * FROM dbo.Customers
SELECT * FROM dbo.OrderLines

And then I have another batch that I run four queries in (three are the same as the other batch), and save the plans to one .sqlplan file, but this is the order:

SELECT * FROM dbo.Orders
SELECT * FROM dbo.Customers
SELECT * FROM dbo.Invoices
SELECT * FROM dbo.OrderLines

I can select which query from each plan I want to display and compare against, correct?

File: src/sql/workbench/api/browser/mainThreadDataProtocol.ts [hunk 1]
```diff
@@ -553,7 +553,8 @@ export class MainThreadDataProtocol extends Disposable implements MainThreadData
 
 	public $registerExecutionPlanProvider(providerId: string, handle: number): void {
 		this._executionPlanService.registerProvider(providerId, <azdata.executionPlan.ExecutionPlanProvider>{
-			getExecutionPlan: (planFile: azdata.executionPlan.ExecutionPlanGraphInfo) => this._proxy.$getExecutionPlan(handle, planFile)
+			getExecutionPlan: (planFile: azdata.executionPlan.ExecutionPlanGraphInfo) => this._proxy.$getExecutionPlan(handle, planFile),
```
Comment (lewis-sanchez) [inline src/sql/workbench/api/browser/mainThreadDataProtocol.ts:556]: What's the meaning behind using a "$"? Is that how we distinguish our API method calls from regular method calls?

**Comment by aasimkhan30** [comment]
> > If you want to open compare two plans in the same file, you will have to add the file twice and then compare the query plans by selecting the right ones in both dropdowns.
> 
> Ah, that wasn't my question. Let's say I have these 3 queries in a batch that I run together, and so there is one .sqlplan file with the plan for all three queries:
> 
> SELECT * FROM dbo.Orders SELECT * FROM dbo.Customers SELECT * FROM dbo.OrderLines
> 
> And then I have another batch that I run four queries in (three are the same as the other batch), and save the plans to one .sqlplan file, but this is the order:
> 
> SELECT * FROM dbo.Orders SELECT * FROM dbo.Customers SELECT * FROM dbo.Invoices SELECT * FROM dbo.OrderLines
> 
> I can select which query from each plan I want to display and compare against, correct?

@erinstellato-ms, yeah that seems correct to me. You should be able to select any  queries from the top and bottom plan files to compare against each other.

**Comment by erinstellato-ms** [comment]
@aasimkhan30 Ok, there are benefits to that.  What I would really love is if it could detect the same query in the other plan and jump to that (e.g. SELECT * FROM dbo.Customers in the baseline plan, and it immediately selects SELECT * FROM dbo.Customers in the second plan).  However, I think that takes a lot more work (you would have to match on something like query_hash), and there can be subtle differences between the queries that might make them look different programmatically.

**Review by kburtram — APPROVED** [review]
LGTM.  A little too much code to do a detailed review, but since there is active iteration in this area & much pending testing it's fine.  Let me know if there are any particular areas you'd like a more detailed review for and I'll take another look.

**Review by lewis-sanchez — APPROVED** [review]

**Pull request merged**